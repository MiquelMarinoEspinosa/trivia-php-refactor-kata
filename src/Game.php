<?php

namespace Game;

final class Game
{
    private const int MAX_NUM_QUESTIONS = 50;

    public private(set) array $popQuestions;
    public private(set) array $scienceQuestions;
    public private(set) array $sportsQuestions;
    public private(set) array $rockQuestions;

    private GameCalculator $gameCalculator;

    public function __construct()
    {
        $this->gameCalculator = new GameCalculator();

        $this->popQuestions = [];
        $this->scienceQuestions = [];
        $this->sportsQuestions = [];
        $this->rockQuestions = [];

        $this->createQuestions();
    }

    public function add(string $playerName): bool
    {
        $this->gameCalculator->add($playerName);

        $this->printAdd($playerName);

        return true;
    }

    public function roll(int $roll): void
    {
        $player = $this->gameCalculator->currentPlayer();

        $this->gameCalculator->roll($roll);

        $this->printRoll($roll, $player);
    }

    public function wasCorrectlyAnswered(): bool
    {
        $player = $this->gameCalculator->currentPlayer();

        $this->gameCalculator->correctAnswer();

        return $this->printAnswerCorrect($player);
    }

    public function wrongAnswer(): bool
    {
        $this->printWrongAnswer();
        $this->gameCalculator->wrongAnswer();
        return true;
    }

    public function isPlayable(): bool
    {
        return $this->gameCalculator->isPlayable();
    }

    private function createQuestions(): void
    {
        for ($numQuestion = 0; $numQuestion < self::MAX_NUM_QUESTIONS; $numQuestion++) {
            $this->createCategoryQuestionsBy($numQuestion);
        }
    }

    private function createCategoryQuestionsBy(int $numQuestion): void
    {
        array_push($this->popQuestions, "Pop Question " . $numQuestion);
        array_push($this->scienceQuestions, "Science Question " . $numQuestion);
        array_push($this->sportsQuestions, "Sports Question " . $numQuestion);
        array_push($this->rockQuestions, "Rock Question " . $numQuestion);
    }

    private function printAdd(string $playerName): void
    {
        $this->echoln($playerName . " was added");
        $this->echoln("They are player number " . $this->gameCalculator->numPlayers());
    }

    private function printRoll(int $roll, int $player): void
    {
        $this->echoln($this->gameCalculator->nameBy($player) . " is the current player");
        $this->echoln("They have rolled a " . $roll);

        $this->printPenaltyBoxMessage($roll, $player);

        if ($this->gameCalculator->isCurrentPlayerGettingOutOfPenaltyBox() === false) {
            return;
        }

        $this->echoln($this->gameCalculator->nameBy(
            $this->gameCalculator->currentPlayer()
        ) . "'s new location is "
          .$this->gameCalculator->currentPlayerPlaces());
        $this->echoln("The category is " . $this->currentCategory());
        $this->askQuestion();
    }

    private function printPenaltyBoxMessage(int $roll, int $player): void
    {
        if (!$this->gameCalculator->isPlayerInPenaltyBox($player)) {
            return;
        }

        $this->echoln($this->gameCalculator->nameBy($player) . $this->buildPenaltyBoxMessage($roll));
    }

    private function buildPenaltyBoxMessage(int $roll): string
    {
        if ($this->gameCalculator->isGettingOutOfPenaltyBoxBy($roll)) {
            return " is getting out of the penalty box";
        }

        return " is not getting out of the penalty box";
    }

    private function printAnswerCorrect(string $player): bool
    {
        if ($this->gameCalculator->isCurrentPlayerGettingOutOfPenaltyBox() === false) {
            return true;
        }

        $this->echoln("Answer was correct!!!!");
        $this->echoln($this->gameCalculator->nameBy($player)
                . " now has "
                . $this->gameCalculator->pursesBy($player)
                . " Gold Coins.");

        return $this->gameCalculator->didPlayerWin($player);
    }

    private function printWrongAnswer(): void
    {
        $this->echoln("Question was incorrectly answered");
        $this->echoln($this->gameCalculator->nameBy(
            $this->gameCalculator->currentPlayer()
        ) . " was sent to the penalty box");
    }

    private function askQuestion(): void
    {
        $this->echoln($this->currentQuestion());
    }

    private function currentQuestion(): string
    {
        $question = match($this->currentCategory()) {
            "Pop"       => array_shift($this->popQuestions),
            "Science"   => array_shift($this->scienceQuestions),
            default     => ""
        };

        if ($this->currentCategory() == "Sports") {
            $question = array_shift($this->sportsQuestions);
        }
        if ($this->currentCategory() == "Rock") {
            $question = array_shift($this->rockQuestions);
        }

        return $question;
    }

    private function currentCategory(): string
    {
        if ($this->gameCalculator->currentPlayerPlaces() == 0) {
            return "Pop";
        }
        if ($this->gameCalculator->currentPlayerPlaces() == 4) {
            return "Pop";
        }
        if ($this->gameCalculator->currentPlayerPlaces() == 8) {
            return "Pop";
        }
        if ($this->gameCalculator->currentPlayerPlaces() == 1) {
            return "Science";
        }
        if ($this->gameCalculator->currentPlayerPlaces() == 5) {
            return "Science";
        }
        if ($this->gameCalculator->currentPlayerPlaces() == 9) {
            return "Science";
        }
        if ($this->gameCalculator->currentPlayerPlaces() == 2) {
            return "Sports";
        }
        if ($this->gameCalculator->currentPlayerPlaces() == 6) {
            return "Sports";
        }
        if ($this->gameCalculator->currentPlayerPlaces() == 10) {
            return "Sports";
        }
        return "Rock";
    }

    private function echoln($string): void
    {
        echo $string."\n";
    }
}
