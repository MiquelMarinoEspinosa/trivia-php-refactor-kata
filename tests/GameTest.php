<?php

declare(strict_types=1);

namespace Tests;

use Game\Game;
use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;

final class GameTest extends TestCase
{
    public function testCreateGame(): void
    {
        ob_start();

        $aGame = new Game();
  
        $aGame->add("Chet");
        $aGame->add("Pat");
        $aGame->add("Sue");

        $aGame->roll(1);
        $aGame->wrongAnswer();

        $aGame->roll(2);
        $aGame->wrongAnswer();

        $aGame->roll(3);
        $aGame->wasCorrectlyAnswered();

        $output = ob_get_clean();

        Approvals::verifyString($output);
    }
}
