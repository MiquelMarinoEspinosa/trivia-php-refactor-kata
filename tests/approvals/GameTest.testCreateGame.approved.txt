Player1 was added
They are player number 1
Player2 was added
They are player number 2
Player3 was added
They are player number 3
Player1 is the current player
They have rolled a 1
Player1's new location is 1
The category is Science
Science Question 0
Question was incorrectly answered
Player1 was sent to the penalty box
Player2 is the current player
They have rolled a 2
Player2's new location is 2
The category is Sports
Sports Question 0
Question was incorrectly answered
Player2 was sent to the penalty box
Player3 is the current player
They have rolled a 3
Player3's new location is 3
The category is Rock
Rock Question 0
Answer was corrent!!!!
Player3 now has 1 Gold Coins.
Player1 is the current player
They have rolled a 4
Player1 is not getting out of the penalty box
Player2 is the current player
They have rolled a 5
Player2 is getting out of the penalty box
Player2's new location is 7
The category is Rock
Rock Question 1
Answer was correct!!!!
Player2 now has 1 Gold Coins.
Player3 is the current player
They have rolled a 6
Player3's new location is 9
The category is Science
Science Question 1
Answer was corrent!!!!
Player3 now has 2 Gold Coins.
Player1 is the current player
They have rolled a 3
Player1 is getting out of the penalty box
Player1's new location is 4
The category is Pop
Pop Question 0
Answer was correct!!!!
Player1 now has 1 Gold Coins.
Player2 is the current player
They have rolled a 6
Player2 is not getting out of the penalty box
Player3 is the current player
They have rolled a 5
Player3's new location is 2
The category is Sports
Sports Question 1
Question was incorrectly answered
Player3 was sent to the penalty box
Player1 is the current player
They have rolled a 4
Player1 is not getting out of the penalty box
Player2 is the current player
They have rolled a 3
Player2 is getting out of the penalty box
Player2's new location is 10
The category is Sports
Sports Question 2
Question was incorrectly answered
Player2 was sent to the penalty box
Player3 is the current player
They have rolled a 2
Player3 is not getting out of the penalty box
Question was incorrectly answered
Player3 was sent to the penalty box
Player1 is the current player
They have rolled a 1
Player1 is getting out of the penalty box
Player1's new location is 5
The category is Science
Science Question 2
Answer was correct!!!!
Player1 now has 2 Gold Coins.
Player2 is the current player
They have rolled a 5
Player2 is getting out of the penalty box
Player2's new location is 3
The category is Rock
Rock Question 2
Question was incorrectly answered
Player2 was sent to the penalty box
Player3 is the current player
They have rolled a 6
Player3 is not getting out of the penalty box
Player1 is the current player
They have rolled a 3
Player1 is getting out of the penalty box
Player1's new location is 8
The category is Pop
Pop Question 1
Question was incorrectly answered
Player1 was sent to the penalty box
Player2 is the current player
They have rolled a 5
Player2 is getting out of the penalty box
Player2's new location is 8
The category is Pop
Pop Question 2
Question was incorrectly answered
Player2 was sent to the penalty box
Player3 is the current player
They have rolled a 2
Player3 is not getting out of the penalty box
Player1 is the current player
They have rolled a 5
Player1 is getting out of the penalty box
Player1's new location is 1
The category is Science
Science Question 3
Question was incorrectly answered
Player1 was sent to the penalty box
Player2 is the current player
They have rolled a 6
Player2 is not getting out of the penalty box
Player3 is the current player
They have rolled a 3
Player3 is getting out of the penalty box
Player3's new location is 5
The category is Science
Science Question 4
Answer was correct!!!!
Player3 now has 3 Gold Coins.
Player1 is the current player
They have rolled a 5
Player1 is getting out of the penalty box
Player1's new location is 6
The category is Sports
Sports Question 3
Answer was correct!!!!
Player1 now has 3 Gold Coins.
Player2 is the current player
They have rolled a 2
Player2 is not getting out of the penalty box
